numpy/core/src/umath/ufunc_object.hpp

```diff
@@ -0,0 +1,292 @@
+        "variant order overloading is not supported yet"
+    );
+    static_assert(
+        ((ufunc_helper::FuncDeduce<TFuncOverloads>::kNumOut == kNumOut_) && ...),
+        "Overloaded functions must have the same number of output,"
+        "variant order overloading is not supported yet"
+    );
+    static_assert(
+        (ufunc_helper::FuncDeduce<TFuncOverloads>::kIsSlice && ...),
+        "Only np::Slice<T> is supported as parmater type at current moment"
+    );
+
+    // concate all types of all functions parmaters into a single tuple
+    using AllParamsTypes_ = decltype(std::tuple_cat(
+        std::declval<typename ufunc_helper::FuncDeduce<TFuncOverloads>::Args>()...
+    ));
+
+    // the main constracing function, requires the indices
+    // of `AllParamsTypes_` so we keep it private.
+    template <std::size_t ...Ind>
+    constexpr UFuncObject(const char *name, const char *doc,
+                          int identity, PyObject *identity_value,
+                          const char *signature, TFuncOverloads ...funcs,
+                          std::index_sequence<Ind...>
+    )
+        : // intilize C function pointers
+          functions_{ToLegacyCFunc_(std::forward<TFuncOverloads>(funcs))...},
+          // convert all parmaters types into TypeIDs
+          ids_{ufunc_helper::ElementTypeID<
+              std::tuple_element_t<Ind, AllParamsTypes_>
+          >...},
+          // Initlize the C++ function pointers
+          data_{reinterpret_cast<void*>(funcs)...},
+          ref_(
+              PyUFunc_FromFuncAndDataAndSignatureAndIdentity(
+                  functions_, data_, ids_, kNumOverloads, kNumIn_, kNumOut_, identity,
+                  name, doc, 0/*unused*/, signature, identity_value
+              )
+          )
+    {}
+
+    template <typename TFunc, size_t ...Ind>
+    static constexpr size_t Index_(std::index_sequence<Ind...>)
+    {
+        static_assert(
+            (std::is_same_v<TFuncOverloads, TFunc> || ...),
+            "Unable to find an existance signature for this overload"
+        );
+        return ((std::is_same_v<TFuncOverloads, TFunc> ? Ind + 1 : 0) + ...) - 1;
+    }
+
+    // returns C function pointer after wraps C++ calls, we use
+    // data* to store C++ function pointer
+    template <typename TFunc, typename TArgs, size_t ...Ind>
+    static constexpr PyUFuncGenericFunction ToLegacyCFunc_(std::index_sequence<Ind...>)
+    {
+        return [](char **args, IntPtr const *dimensions, IntPtr const *strides, void *data) -> void {
+            using base_types = std::tuple<std::remove_reference_t<
+                std::tuple_element_t<Ind, TArgs>
+            >...>;
+            auto params = base_types(
+                std::tuple_element_t<Ind, base_types>(
+                    args[Ind], dimensions[0], strides[Ind]
+                )...
+            );
+            std::apply(reinterpret_cast<TFunc>(data), params);
+        };
```

Comment:
So, I think my main worry is this part.  We already have two loop versions:
* The legacy that is unwrapped here.
* A new loop signature, which is largely the same (but adds more info and an error return).  This is not accessible through `PyUFunc_FromFuncAndDataAndSignatureAndIdentity`, but we should aim to replace it sooner rather than later internally.
  * This signature is very much open to change, but I would like to be sure that updating existing code is easy eventually.
  * We don't *need* a "legacy" loop, we only have to support it when downstream gives us one.
  * This new-loop can be specialized for specific strides (in most ufunc calls) for speed, so that selection of SIMD can happen outside of the inner-loop itself.

Since you use the static `void *data`, the compiler cannot possibly optimize this away either.  The actual (new) ufunc signature currently doesn't even include static data (but we can add that back in some form, it is likely useful also for generic heavy functions, where we currently just pass a function pointer also).

Note that one thing I would like mid-term, is to move the current `if (stride[0] == 0)`, etc. branches out of the inner-loop.
The machinery exists, but we have to start using it.  This refactor should help with that since it should make our ufunc generation just a bit less of a mess hopefully, and to get there we should update most/all ufunc signatures to the new-style and adapt the ufunc generation a fair bit.

---

I guess you want C++ iterators and I expect they are really helpful and maybe necessary for later SIMD machinery improvements since those make iteration more painful.

But I do need to ask what the alternatives are?  Would it be really be terrible to build an iterator (or multiple iterators) _inside_ the inner-loop with a one-line per-argument boilerplate?  That way you dog-feed the C-signature _explicitly_ but still iterate in whatever way we like.  Also, I do think it would make it much clearer to have both code styles side-by-side without getting super confused.  And we will always have both side-by-side, because the C-signature is still our public API (unless we have the guts to change that, which may be a path, but probably a more difficult and slower one than I would like right now).

---

One more question I have but not for Sayed explicitly.  I suspected we would aim to just remove the Python generation completely, even if we end up with a rather long C++ file (that could be generated via an adapted old file once for reference).

But to me, that would be another advantage of starting of with not modifying the signature at all, because then we can do the two steps separately. (Yes, this requires a bit of double work, but on the plus side you can do it in chunks.)

Reply:
> A new loop signature, which is largely the same (but adds more info and an error return).

`PyArrayMethod_Context`, and `int` rather then `void` as return, well not sure how useful `PyArrayMethod_Context` is but it can be supported through C++ also new C++ signature can support either `bool` or `int` for handling errors even if I think that testing `PyErr_Occurred()` is quite enough or maybe we can enable C++ exception in some stages.

> ` This is not accessible through PyUFunc_FromFuncAndDataAndSignatureAndIdentity, but we should aim to replace it sooner rather than later internally.`

If we are moving to `C++` then what the point of continue enhancing C interface? Building a new interface in C++ to count more on static optimization should significantly speed up the internal calls.

> This signature is very much open to change, but I would like to be sure that updating existing code is easy eventually.
  
The current approach provides the best start for the highest layer but it costs extra later for sure we can enhance it later at least by cutting off the legacy layer.

> This new-loop can be specialized for specific strides (in most ufunc calls) for speed, so that selection of SIMD can happen outside of the inner-loop itself.

I'm not sure how it's going to be useful for SIMD kernels, we usually support both contiguous and non-contiguous memory access as long as there's no memory overlapping. Specialize within the inner loop is much easier and can save binary size too, For example, we can easily implement a C++ meta class or function that takes two operators one for scalar operation and another for SIMD intrinsic then it's generating all routined branches including SIMD vector memory access, unrolling, reduction and scalar fallback.

> Since you use the static void *data, the compiler cannot possibly optimize this away either.

I don't think it's necessary to support it for C++, there're better patterns.

> Note that one thing I would like mid-term, is to move the current if (stride[0] == 0), etc. branches out of the inner-loop.

If counting on meta functions to generate these branches wouldn't be robust then we can discuss moving to specialize these branches through UFunc API, after this pr gets merged I will provide real examples in this matter.

> I guess you want C++ iterators and I expect they are really helpful and maybe necessary for later SIMD machinery improvements since those make iteration more painful.

It is also important for type safety.

> But I do need to ask what the alternatives are?
That way you dog-feed the C-signature explicitly but still iterate in whatever way we like.

Avoid using C++ functions pointers and wraps the C UFunc API through meta programming, this solution is zero-cost overhead but unfortunately not that flexible and increases the complexity of the code especially when exporting internal symbols for CPU runtime dispatching.

>  And we will always have both side-by-side, because the C-signature is still our public API (unless we have the guts to change that, which may be a path, but probably a more difficult and slower one than I would like right now).

I agree but nothing is perfect, maybe provides inline functions that can serve both directions when it's needs until C++ dominates.
> suspected we would aim to just remove the Python generation

Yes, but we also have to get rid of C template sources too. all at once seems more convenient to me. The new signature works only internally and can be improved later at any time.

Reply:
I made changes to the code, to remove the runtime overhead and pass the inner loops as objects rather than function pointers (fully meta) please take a look at commit https://github.com/numpy/numpy/pull/19753/commits/196949481c915e844bde9b142b979a72986953a2 to realize the differences:

However, once we deal with runtime dispatch or any exported symbols the overhead will come back again. I still prefer the previous method.

Reply:
Another suggestion to C++ signature is to change the output indicator from reference to non-const:
```diff
-(Slice<T> in0, Slice<T> in1, Slice<T> &out)
+(const Slice<T> in0, const Slice<T> in1, Slice<T> out) 
```

The number of inputs and outputs are deduced from the signature rather than specifying it manually in case we decided later to support non-static signatures similar to function overloading in C++.

Reply:
I won't have time to look at this for a few weeks.  But since Marten commented: Yes, we should revive this.

FWIW, I am still a bit unsure about the right choice of inner-loop.  Being more used to C++ these days helps, though.

My main question I guess remains a bit:  I think you wrote this under the assumption to make the full thing C++.  I lack a bit of clarity that this is a good choice, since we *must* keep a first class public C-API.

I suppose I can rephrase this into:  Please assume that you will *never* be able to remove the step to translate the current C call into a C++ internal call.  Would you reconsider this API choice?  If not, then probably fine.  But if yes, what would it look like? 
(Even if it is just something creating a `MultiIter<...>`, whether more complicated or not to do the C/C++ layer conversion.)

(I would have been very happy, and still would be happy, to tweak the C signature.  But moving where the C/C++ interface remains a big unclear thing to me that I don't know if it could or should be done.  If others are more clear on it, that would be fine...)